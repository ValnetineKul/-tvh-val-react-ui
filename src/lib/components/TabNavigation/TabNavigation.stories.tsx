import type { ComponentProps } from 'react';
import React from 'react';
import { createMeta, createTemplate } from '../../story-utils';
import ChipGroup from '../Chips/ChipGroup';
import ChoiceChip from '../Chips/ChoiceChip';
import Typography from '../Typography';
import TabNavigation from './TabNavigation';

type Props = ComponentProps<typeof TabNavigation>;

const Template = createTemplate(TabNavigation);

const chips = ['Case-IH', 'Claas/Renault', 'Deutz', 'Ford', 'JCB'];

const tabs: Props['tabList'] = [
  {
    id: 'Text',
    label: 'Text',
    component: 'Simple text',
  },
  {
    id: 'React component',
    label: 'React component',
    count: 1,
    component: (
      <div>
        <Typography variant="h4">Title</Typography>
        <Typography>Another React component is rendered in the tab</Typography>
      </div>
    ),
  },
  {
    id: 'Empty tab',
    label: 'Empty tab',
    component: null,
  },
  {
    id: 'Brands',
    label: 'Brands',
    count: 99,
    component: (
      <ChipGroup>
        {chips.map((chip, index) => {
          return <ChoiceChip label={chip} key={index} />;
        })}
      </ChipGroup>
    ),
  },
];
for (let i = 4; i < 7; i++) {
  const label = `Tab ${i + 1}`;
  const component = `Tab ${i + 1} content`;

  tabs.push({
    label,
    component,
    id: label,
  });
}

export const Primitive = Template.bind({});
Primitive.args = {
  tabList: tabs,
  activeTab: 'React component',
};
Primitive.storyName = 'TabNavigation';

export default createMeta({
  component: TabNavigation,
  title: 'Components/TabNavigations/TabNavigation',
  parameters: {
    backgrounds: {
      default: 'theme1',
      values: [
        { name: 'theme0', value: '#fff' },
        { name: 'theme1', value: '#E6E6E6' },
      ],
    },
  },
  argTypes: {
    tabList: {
      control: {
        disable: true,
      },
    },
  },
});
