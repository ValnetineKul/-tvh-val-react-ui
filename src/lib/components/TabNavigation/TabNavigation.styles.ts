import { makeStyles } from '../../../themes/core';

const useStyles = makeStyles<void, 'scrollButtons'>({ name: 'TabNavigation' })((theme, _params, classes) => ({
  root: {
    minHeight: 'unset',
    [`& .${classes.scrollButtons}`]: {
      display: 'inline-flex',
    },
  },

  tab: {
    color: theme.color['text/onSurface100/default'],
    backgroundColor: theme.color['bg/surfaceCurrent/default'],
    opacity: 1,
    textAlign: 'left',
    textTransform: 'none',
    minWidth: 'unset',
    minHeight: 'unset',
    padding: `${theme.layoutSpacing['spacing/300']} ${theme.layoutSpacing['spacing/400']}`,
    position: 'relative',

    '&[aria-selected=true]': {
      zIndex: 1,
    },

    '&:focus': {
      color: theme.color['text/onSurface100/default'],
      '&::after': {
        content: '""',
        display: 'block',
        position: 'absolute',
        left: 0,
        top: 0,
        width: '100%',
        height: '100%',
        boxShadow: `inset 0px 0px 0px 4px ${theme.color['border/onSurface100/focus']}`,
      },
    },

    '&:hover': {
      backgroundColor: theme.color['bg/surface100/hover'],
    },

    '&:active': {
      backgroundColor: theme.color['bg/surface100/active'],
    },
  },

  tabSelected: {
    '&.Mui-selected': {
      color: theme.color['text/onSurface100/default'],
      backgroundColor: theme.color['bg/surface100/default'],

      '&:hover': {
        backgroundColor: theme.color['bg/surface100/hover'],
      },

      '&:active': {
        backgroundColor: theme.color['bg/surface100/active'],
      },
    },

    '&:before': {
      content: '""',
      width: '100%',
      height: 2,
      backgroundColor: theme.color['border/onSurfaceCurrent/selected'],
      position: 'absolute',
      left: 0,
      top: 0,
    },
  },

  tabContent: {
    padding: theme.layoutSpacing['spacing/400'],
    position: 'relative',

    '&:focus': {
      outline: 'none',
      boxShadow: `0px 0px 0px 4px ${theme.color['border/onSurface100/focus']}`,
    },
  },

  tabContainer: {
    position: 'relative',
  },

  indicator: {
    display: 'none',
  },

  bottomDivider: {
    position: 'absolute',
    width: '100%',
    bottom: 0,
  },

  verticalDivider: {
    position: 'absolute',
    bottom: 0,
    height: 'calc(100% - 2px)',
  },

  verticalDividerLeft: {
    left: 0,
  },

  verticalDividerRight: {
    right: 0,
  },

  scrollButtons: {},
}));

export default useStyles;
