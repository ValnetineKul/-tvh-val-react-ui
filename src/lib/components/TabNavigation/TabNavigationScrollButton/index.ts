import TabNavigationScrollButton from './TabNavigationScrollButton';

export type { TabNavigationScrollButtonProps } from './TabNavigationScrollButton';
export default TabNavigationScrollButton;
