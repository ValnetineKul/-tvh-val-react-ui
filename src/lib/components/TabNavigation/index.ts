import TabNavigation from './TabNavigation';

export type { TabNavigationProps } from './TabNavigation';
export default TabNavigation;
