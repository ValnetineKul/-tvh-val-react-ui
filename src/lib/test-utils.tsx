import userEvent from '@testing-library/user-event';
import { fireEvent, render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { theme } from '../themes/tvh';
import { createTheme, ThemeProvider, SurfaceProvider } from '../themes/core';

export const muiTheme = createTheme(theme);

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <MemoryRouter>
      <ThemeProvider theme={muiTheme}>
        <SurfaceProvider color="100">{children}</SurfaceProvider>
      </ThemeProvider>
    </MemoryRouter>
  );
};

const customRender = (component: React.ReactElement, options?: Parameters<typeof render>[1]) =>
  render(component, { wrapper: AllTheProviders, ...options });

const customRenderHook: typeof renderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

export const initResponsiveTest = (size: number | 'Mobile' | 'Tablet' | 'Desktop' | 'LgDesktop' | 'XlDesktop') => {
  let width: number;
  if (size === 'Mobile') width = muiTheme.breakpoints.values.sm - 1;
  else if (size === 'Tablet') width = muiTheme.breakpoints.values.sm;
  else if (size === 'Desktop') width = muiTheme.breakpoints.values.md;
  else if (size === 'LgDesktop') width = muiTheme.breakpoints.values.lg;
  else if (size === 'XlDesktop') width = muiTheme.breakpoints.values.xl;
  else width = size;

  Object.defineProperty(window, 'innerWidth', {
    writable: true,
    configurable: true,
    value: width,
  });

  window.matchMedia = jest.fn().mockImplementation((query) => {
    const queryWidthInterval: string[] = query.match(/\d+(\.\d+)?/g);

    let matches;
    if (queryWidthInterval.length === 2) {
      matches = width >= +queryWidthInterval[0] && width < +queryWidthInterval[1];
    } else {
      const queryWidth = +queryWidthInterval[0];
      const isDown = /max/.test(query);

      matches = isDown ? width <= queryWidth : width >= queryWidth;
    }

    return {
      matches,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
    };
  });
};

const customUserEvent = Object.assign(userEvent, {
  mouseDown: fireEvent.mouseDown,
  mouseUp: fireEvent.mouseUp,
  mouseMove: fireEvent.mouseMove,
  scroll: fireEvent.scroll,
  error: fireEvent.error,
});

export { screen, within, waitFor, act, waitForElementToBeRemoved } from '@testing-library/react';
// override render method
export { customRender as render, customRenderHook as renderHook, customUserEvent as userEvent };
