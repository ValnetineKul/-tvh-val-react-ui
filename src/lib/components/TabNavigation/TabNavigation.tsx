import { default as MuiTabItem } from '@mui/material/Tab';
import { default as MuiTabs } from '@mui/material/Tabs';
import type { FC, ReactNode } from 'react';
import React, { useEffect, useState } from 'react';
import Typography from '../Typography';
import Surface from '../Surfaces/Surface';
import Divider from '../Divider';
import Badge from '../Badge';
import TabNavigationScrollButton from './TabNavigationScrollButton';
import type { DataAttributes } from '../../types/common';
import useStyles from './TabNavigation.styles';

interface TabListItem {
  id: string;
  label: ReactNode;
  component: ReactNode;
  count?: number;
  itemProps?: DataAttributes;
}

export interface TabNavigationProps {
  tabList: TabListItem[];
  activeTab?: string;
  onTabChange?: (id: string) => void;
  className?: string;
  contentClassName?: string;
  tabClassName?: string;
}

const TabNavigation: FC<TabNavigationProps> = ({
  tabList,
  className,
  activeTab,
  onTabChange,
  contentClassName,
  tabClassName,
}) => {
  const { classes, cx } = useStyles();
  const [value, setValue] = useState(0);

  useEffect(() => {
    if (!!activeTab && typeof activeTab === 'string') {
      const index = tabList.findIndex((tab) => tab.id === activeTab);
      if (index === -1) return;
      setValue(index);
    }
  }, [activeTab, tabList]);

  const handleChange = (event: React.ChangeEvent, newValue: number) => {
    if (tabList[newValue] && onTabChange) {
      onTabChange(tabList[newValue].id);
    }

    setValue(newValue);
  };

  const renderLabel = (label: ReactNode, isSelected: boolean) => {
    return (
      <Typography component="span" weight={isSelected ? ('emphasis' as const) : ('regular' as const)}>
        {label}
      </Typography>
    );
  };

  const renderTabItems = ({ label, count, itemProps }: TabListItem, key: number) => {
    const isSelected = value === key;

    return (
      <MuiTabItem
        id={`tabItem-${key}`}
        aria-controls={`tabContent-${key}`}
        key={key}
        label={
          <>
            {isSelected && (
              <Divider direction="vertical" className={cx(classes.verticalDivider, classes.verticalDividerLeft)} />
            )}
            {count ? (
              <Badge count={count || 0} variant="default">
                {renderLabel(label, isSelected)}
              </Badge>
            ) : (
              renderLabel(label, isSelected)
            )}
            {isSelected && (
              <Divider direction="vertical" className={cx(classes.verticalDivider, classes.verticalDividerRight)} />
            )}
          </>
        }
        classes={{
          root: cx(classes.tab, tabClassName),
          selected: classes.tabSelected,
        }}
        {...itemProps}
      />
    );
  };

  const renderTabContent = ({ component }: TabListItem, key: number) => {
    return (
      <div
        id={`tabContent-${key}`}
        key={key}
        className={cx(classes.tabContent, contentClassName)}
        role="tabpanel"
        aria-labelledby={`tabItem-${key}`}
        hidden={value !== key}
        tabIndex={0}
      >
        {component}
      </div>
    );
  };

  return (
    <div className={className}>
      <div className={classes.tabContainer}>
        <MuiTabs
          value={value}
          onChange={handleChange}
          classes={{
            root: classes.root,
            indicator: classes.indicator,
            scrollButtons: classes.scrollButtons,
          }}
          variant="scrollable"
          scrollButtons="auto"
          ScrollButtonComponent={({ orientation, ...restProps }) => <TabNavigationScrollButton {...restProps} />}
        >
          {tabList.map(renderTabItems)}
        </MuiTabs>
        <Divider className={classes.bottomDivider} />
      </div>
      <Surface color="100">{tabList.map(renderTabContent)}</Surface>
    </div>
  );
};

export default TabNavigation;
