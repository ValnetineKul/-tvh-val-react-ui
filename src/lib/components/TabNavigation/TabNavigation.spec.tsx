import React from 'react';
import { render } from '../../test-utils';
import TabNavigation from './TabNavigation';

const tabs = [
  {
    id: 'First tab',
    label: 'First tab',
    component: 'First tab content',
  },
  {
    id: 'Second tab',
    label: 'Second tab',
    count: 12,
    component: 'Second tab content',
  },
  {
    id: 'Third tab',
    label: 'Third tab',
    component: 'Third tab content',
  },
];

describe('TabNavigation', () => {
  it('Should render correctly', () => {
    const { container } = render(<TabNavigation tabList={tabs} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Should add custom className', () => {
    const { container } = render(<TabNavigation tabList={tabs} className="custom-class-name" />);
    expect(container.firstChild).toHaveClass('custom-class-name');
  });

  it('Should render badge', () => {
    const { getByText } = render(<TabNavigation tabList={tabs} />);
    const badge = getByText(/12/);
    expect(badge).toBeInTheDocument();
  });

  it('Should switch tab on click', async () => {
    const onTabChange = jest.fn();
    const { getByRole } = render(<TabNavigation tabList={tabs} onTabChange={onTabChange} />);
    const firstTab = getByRole('tab', { name: 'First tab' });
    const secondTab = getByRole('tab', { name: 'Second tab 12' });

    expect(getByRole('tabpanel').textContent).toMatch('First tab content');

    secondTab.click();
    expect(getByRole('tabpanel').textContent).toMatch('Second tab content');
    expect(onTabChange).toHaveBeenCalledTimes(1);
    expect(onTabChange).toHaveBeenCalledWith('Second tab');

    firstTab.click();
    expect(getByRole('tabpanel').textContent).toMatch('First tab content');
    expect(onTabChange).toHaveBeenCalledTimes(2);
    expect(onTabChange).toHaveBeenCalledWith('First tab');
  });

  it('Should set active tab', () => {
    const { getByText } = render(<TabNavigation tabList={tabs} activeTab="Second tab" />);
    const activeButtonTab = getByText('Second tab').closest('button');
    expect(activeButtonTab).toHaveAttribute('aria-selected', 'true');
  });
});
