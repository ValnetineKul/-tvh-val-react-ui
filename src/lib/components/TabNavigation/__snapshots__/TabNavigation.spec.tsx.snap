// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabNavigation Should render correctly 1`] = `
<div>
  <div
    class="tss-1trw2ln-TabNavigation-tabContainer"
  >
    <div
      class="tss-v8c4vf-TabNavigation-root MuiTabs-root mui-1ujnqem-MuiTabs-root"
    >
      <div
        class="MuiTabs-scrollableX MuiTabs-hideScrollbar mui-agdkjh-MuiTabs-scrollbarSize"
        style="width: 99px; height: 99px; position: absolute; top: -9999px; overflow: scroll;"
      />
      <div
        class="MuiTabs-scroller MuiTabs-hideScrollbar MuiTabs-scrollableX mui-69z67c-MuiTabs-scroller"
        style="margin-bottom: -100px;"
      >
        <div
          class="MuiTabs-flexContainer mui-heg063-MuiTabs-flexContainer"
          role="tablist"
        >
          <button
            aria-controls="tabContent-0"
            aria-selected="true"
            class="MuiButtonBase-root tss-1i7euy5-TabNavigation-tab MuiTab-root MuiTab-textColorPrimary tss-1mg7piw-TabNavigation-tabSelected Mui-selected mui-18etwwi-MuiButtonBase-root-MuiTab-root"
            id="tabItem-0"
            role="tab"
            tabindex="0"
            type="button"
          >
            <hr
              class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem tss-1utnvid-Divider-root-TabNavigation-verticalDivider-TabNavigation-verticalDividerLeft-Divider-vertical mui-1tk1bks-MuiDivider-root"
            />
            <span
              class="tss-0-Typography-weightEmphasis-ref tss-1g15zls-Typography-root-Typography-weightEmphasis-Typography-body500"
            >
              First tab
            </span>
            <hr
              class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem tss-1myoqk5-Divider-root-TabNavigation-verticalDivider-TabNavigation-verticalDividerRight-Divider-vertical mui-1tk1bks-MuiDivider-root"
            />
          </button>
          <button
            aria-controls="tabContent-1"
            aria-selected="false"
            class="MuiButtonBase-root tss-1i7euy5-TabNavigation-tab MuiTab-root MuiTab-textColorPrimary mui-18etwwi-MuiButtonBase-root-MuiTab-root"
            id="tabItem-1"
            role="tab"
            tabindex="-1"
            type="button"
          >
            <div
              class="tss-qicf1j-Badge-root"
            >
              <span
                class="tss-0-Typography-weightRegular-ref tss-6tt2d9-Typography-root-Typography-weightRegular-Typography-body500"
              >
                Second tab
              </span>
              <span
                class="tss-k1vbve-Surface-root-Badge-badge-Badge-badge-Badge-textBadgePosition"
              >
                <span
                  class="tss-0-Typography-weightEmphasis-ref tss-a8jd6j-Typography-root-Typography-weightEmphasis-Badge-badgeTypography-Typography-body300"
                >
                  12
                </span>
              </span>
            </div>
          </button>
          <button
            aria-controls="tabContent-2"
            aria-selected="false"
            class="MuiButtonBase-root tss-1i7euy5-TabNavigation-tab MuiTab-root MuiTab-textColorPrimary mui-18etwwi-MuiButtonBase-root-MuiTab-root"
            id="tabItem-2"
            role="tab"
            tabindex="-1"
            type="button"
          >
            <span
              class="tss-0-Typography-weightRegular-ref tss-6tt2d9-Typography-root-Typography-weightRegular-Typography-body500"
            >
              Third tab
            </span>
          </button>
        </div>
        <span
          class="tss-cxrj87-TabNavigation-indicator MuiTabs-indicator mui-1aquho2-MuiTabs-indicator"
          style="left: 0px; width: 0px;"
        />
      </div>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth tss-1l8gu62-Divider-root-TabNavigation-bottomDivider-Divider-horizontal mui-9mgopn-MuiDivider-root"
    />
  </div>
  <div
    class="tss-bdw984-Surface-root"
  >
    <div
      aria-labelledby="tabItem-0"
      class="tss-1d3h4oy-TabNavigation-tabContent"
      id="tabContent-0"
      role="tabpanel"
      tabindex="0"
    >
      First tab content
    </div>
    <div
      aria-labelledby="tabItem-1"
      class="tss-1d3h4oy-TabNavigation-tabContent"
      hidden=""
      id="tabContent-1"
      role="tabpanel"
      tabindex="0"
    >
      Second tab content
    </div>
    <div
      aria-labelledby="tabItem-2"
      class="tss-1d3h4oy-TabNavigation-tabContent"
      hidden=""
      id="tabContent-2"
      role="tabpanel"
      tabindex="0"
    >
      Third tab content
    </div>
  </div>
</div>
`;
